rfcs/001_context.md

```diff
@@ -0,0 +1,145 @@
+- Feature name: `Cancelable operations and use of Context`
+- Date: 2020-10-10
+- Author: Asdine El Hrychy
+- Related discussions: [Issue #224](https://github.com/genjidb/genji/issues/224), [PR #231](https://github.com/genjidb/genji/pull/231)
+
+## General approach
+
+I would like to avoid context pollution as much as possible, but without restricting user flexibility. It is important to be able to control and cancel operations but I also believe that there are times where we don't care about that.
+I want Genji to be simple to use and intuitive, and for that, I prefer to expand a bit more the surface API, rather than making the existing ones more complete and complex.
+The purpose of this RFC is not to describe in detail how each operation should be canceled but rather which APIs should expect a context parameter.
+
+## Genji package and database lifecycle
+
+### Open and New
+
+Opening a database relies on IO and should be cancelable, but after that operation is completed the database handle itself is long-lived and should not be cancelable, but closable with the `Close()` method.
+However, most of the time, we don't care about how long it takes.
+
+We currently have two ways of opening a database:
+
+- `genji.Open`
+- `genji.New`
+
+`Open` is just a specialized way of opening a database which calls `New`. It is by nature incomplete:
+
+- It only allows opening an in-memory database or a Bolt database
+- It doesn't take any option
+
+```go
+db, err := genji.Open("foo.db")
+if err != nil {
+    return err
+}
+defer db.Close()
+```
+
+It is very simple to write and to use, it must remain as-is.
+
+If users want more power, they must use the `genji.New` function.
+
+Opening a database can take time, even for Bolt or Badger, because accessing disks is unbounded (slow disks, NFS, etc.). It makes sense to have this a cancelable operation.
+
+```go
+ctx, cancel := context.WithTimeout(context.Background(), 5 * time.Second)
+...
+
+// ctx here is only used to control how long the engine creation can take
+// as engines might wait for IO
+ng := someengine.New(ctx, ...)
+// but not all of them
+ng := memoryengine.New()
+
+// at startup, Genji performs some operations on the engine, this should be cancelable
+db, err := genji.New(ctx, ng)
+```
+
+Storing the context for further uses prevents doing that and is not intuitive for a database handle that is supposed to be long-lived.
+
+### genji.DB methods
+
+Each of these must be cancelable independently, at any given time while the database handle is opened.
+Because cancelation is handled independently from the database handle, it is very intuitive and doesn't require preparing a different handle to be able to cancel a query.
+
+```go
+db.Exec(ctx, "INSERT INTO foo(a) VALUES (1)")
```

Comment:
I think we can reuse the context from `genji.Open`, and allow changing it with `WithContext` method. The shallow copy is cheap since `genji.DB` just wraps `database.Database`, same for the `genji.Tx`.

In fact, that’s what #231 already does (i.e. you only need explicit context on `Open`).

Reply:
Reusing that context can cause several issues:

- It is less flexible: We can't control how long opening a database can take since the context is used for the whole database lifecycle, which is not supposed to be canceled very often.
- It is less intuitive: It is way more common to pass a context for the current operation rather than for the lifecycle of a database handle, which again is supposed to be long-lived.
- It makes the API more complex: We must remember to call `WithContext` every time we want one particular action to be cancelable, which is the most common use case. In the context of an http handler for example, we would have to always use `WithContext(req.Context())`. It seems easier to run `db.Exec(req.Context())` or `db.Begin(req.Context(), true)` instead.

If we want to automatically close the database when the root context is canceled, I believe there are simpler ways to do so

Reply:
On the other hand, it’s more flexible because we can pass context to individual operations, and ignore it in case we don’t need it (e.g. command-line apps).

I don’t think it’s that that uncommon, a few examples I can think of right now are
- [http.Request.WithContext](https://golang.org/pkg/net/http#Request.WithContext)
- [gorm.DB.WithContext](https://pkg.go.dev/gorm.io/gorm#DB.WithContext)

It’s also nice to prototype some APIs without context cruft, and then gradually add contexts/cancellation where needed. We’ve been using `gorm` at my last $job, and I’d say that it’s less intrusive than `database/sql` interface.

Even if you missed the context, the change is one line:
```
tx = tx.WithContext(ctx)
// or
db = db.WithContext(ctx)
```

That said, I think that context-per-operation goes against the primary goal to avoid context pollution.

Reply:
One more objection for context per transaction is that we may want to set a timeout on an individual query, but not the whole transaction.

Reply:
Ok so there are two subjects here:

1/ Whether we should reuse the context given during database opening
2/ Whether `db.Exec` and `db.Query*` should receive a context or not

My primary concern with 1/ is that the context should only be used for controlling database opening, and as proposed in this RFC, only passed to `genji.New`.

Regarding 2/ though, I think this is a good idea that can indeed prevent context pollution even further:

```go
db, _ := genji.Open("foo.db")
defer db.Close()

db.Exec("INSERT INTO foo (a) VALUES (1)")

db.WithContext(ctx).Exec("INSERT INTO foo (a) VALUES (1)")

dbx := db.WithContext(ctx)

dbx.Query("SELECT 1")

tx, err := dbx.Begin(true)

dbx.View(func() {...})
```

Reply:
>My primary concern with 1/ is that the context should only be used for controlling database opening, and as proposed in this RFC, only passed to genji.New.

We can address this by defaulting to `context.Background()` instead of the context passed to `Open`.

Reply:
> We can address this by defaulting to context.Background() instead of the context passed to Open.

Exactly

Reply:
> One more objection to the context per transaction is that we (as users) may want to set a timeout on an individual query, but not the whole transaction.

I thought about this but I have trouble seeing a valid use case that would be worth the trouble, i.e. when someone would need to cancel a specific query AND keep running other queries if that one was canceled.